### PR TITLE
[CTRL] Support attn_implementation dispatch

### PR DESCRIPTION
**Pull request by YangKai0616**
# What does this PR do?

The `CTRL` model fails during the initialization phase when using `from_pretrained(..., attn_implementation="sdpa")`. This PR enables standard `attn_implementation="sdpa"` dispatch to `CTRL` .

Hi @ArthurZucker, pls help review, thx!

### CODE REVIEW

**Comment by YangKai0616** [comment]
Hi @vasqu , please help review this PR, thank you!

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 1]
```diff
@@ -54,42 +57,41 @@ def positional_encoding(position, d_model_size, dtype):
...
-        self.num_heads = num_heads
-        self.d_model_size = d_model_size
+        self.config = config
+        self.num_heads = config.n_head
+        self.d_model_size = config.n_embd
         self.layer_idx = layer_idx
+        self.is_causal = True
 
-        self.depth = int(d_model_size / self.num_heads)
+        self.depth = int(self.d_model_size / self.num_heads)
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: I think this model is super old, would rather align a bit more with modern terminology - ig this should be `head_dim`

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: Imo we should remove this function and just follow what llama did (with the same reshapes etc)

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 2]
```diff
@@ -133,40 +143,34 @@ def point_wise_feed_forward_network(d_model_size, dff):
...
-        mask,
         layer_past=None,
         attention_mask=None,
-        use_cache=False,
-        output_attentions=False,
         **kwargs,
     ):
         normed = self.layernorm1(x)
-        attn_outputs = self.multi_head_attention(
+        attn_output = self.multi_head_attention(
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]:
```suggestion
        attn_output, _ = self.multi_head_attention(
```
would rather do this then the `...[0]`

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 3]
```diff
@@ -175,14 +179,18 @@ def forward(
...
 
 
 @auto_docstring
 class CTRLPreTrainedModel(PreTrainedModel):
     config: CTRLConfig
     base_model_prefix = "transformer"
+    _supports_sdpa = True
+    _can_record_outputs = {
+        "hidden_states": EncoderLayer,
+        "attentions": OutputRecorder(MultiHeadAttention, index=1),
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]:
```suggestion
        "attentions": MultiHeadAttention,
```
pretty sure we default to index 1 on attention so no need but not 100% sure

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 4]
```diff
@@ -175,14 +179,18 @@ def forward(
...
-        outputs = (out2,) + attn_outputs[1:]
-        return outputs
+        return out2
 
 
 @auto_docstring
 class CTRLPreTrainedModel(PreTrainedModel):
     config: CTRLConfig
     base_model_prefix = "transformer"
+    _supports_sdpa = True
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py:199]: The attention looks fairly simple, could we also support all the other flags for this - flex, flash, attention backend (needs interface + kwargs passing from top to bottom)?

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 5]
```diff
@@ -224,6 +227,7 @@ def get_input_embeddings(self):
     def set_input_embeddings(self, new_embeddings):
         self.w = new_embeddings
 
+    @capture_outputs
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py:243]: missing merge config with defaults

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 6]
```diff
@@ -54,42 +57,41 @@ def positional_encoding(position, d_model_size, dtype):
     return pos_encoding
 
 
-def scaled_dot_product_attention(q, k, v, mask, attention_mask=None):
-    # calculate attention
-    matmul_qk = torch.matmul(q, k.permute(0, 1, 3, 2))
+def eager_attention_forward(module, query, key, value, attention_mask, scaling=None, dropout=0.0, **kwargs):
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: Imo, we could use a copied from here?

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 7]
```diff
@@ -289,26 +285,6 @@ def forward(
             position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
             position_ids = position_ids.unsqueeze(0)
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py:295]: Not on you but imo we can refactor this as well, see e.g. llama where we create the input embeds earlier and get the shape from there then most of this can be reduced by quite a bit

**Review by vasqu — COMMENTED** [review]
Overall looks good, would like to align a few smaller things here and there but shouldn't be too much. nice work!

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 1 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 2 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 3 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 4 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py:199]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 5 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py:243]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 6 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 7 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py:295]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 8]
```diff
@@ -54,119 +62,124 @@ def positional_encoding(position, d_model_size, dtype):
...
+        self.dense = nn.Linear(self.d_model_size, self.d_model_size)
 
     def forward(
         self,
-        v,
-        k,
-        q,
-        mask,
+        value,
+        key,
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: Lets not change the naming here just to avoid being breaking

Comment (vasqu) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Or use our deprecate (kwarg) decorator to give us some leeway to deprecate

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 9]
```diff
@@ -54,119 +62,124 @@ def positional_encoding(position, d_model_size, dtype):
...
-        q = self.Wq(q)
-        k = self.Wk(k)
-        v = self.Wv(v)
+        input_shape = query.shape[:-1]
+        hidden_shape = (*input_shape, -1, self.head_dim)
 
-        q = self.split_into_heads(q, batch_size)
-        k = self.split_into_heads(k, batch_size)
-        v = self.split_into_heads(v, batch_size)
+        query = self.Wq(query).view(hidden_shape).transpose(1, 2)
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: Super nit but lets do `_states` as in llama (e.g. `query_states`)

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 10]
```diff
@@ -54,119 +62,124 @@ def positional_encoding(position, d_model_size, dtype):
...
 
-        self.dropout1 = nn.Dropout(rate)
-        self.dropout2 = nn.Dropout(rate)
+        self.dropout1 = nn.Dropout(config.resid_pdrop)
+        self.dropout2 = nn.Dropout(config.resid_pdrop)
 
     def forward(
         self,
         x,
-        mask,
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py:151]: Mask is not used, correct so here its fine to just delete?

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 11]
```diff
@@ -259,27 +273,19 @@ def forward(
...
-        output_hidden_states = (
-            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
-        )
-        return_dict = return_dict if return_dict is not None else self.config.return_dict
 
         if input_ids is not None and inputs_embeds is not None:
             raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
-        elif input_ids is not None:
+        if input_ids is not None:
             self.warn_if_padding_and_no_attention_mask(input_ids, attention_mask)
```
Comment (vasqu) [inline src/transformers/models/ctrl/modeling_ctrl.py]: Let's remove this warning, its not super useful and some old relic. And it introduces graph breaks iirc

**Review by vasqu — APPROVED** [review]
Have left a few smaller comments. We could further refactor to also have the same namings by using the conversion mapping but tbh, I don't think it's worth it for this model as it's very old and likely to be deprecated at some point

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 8 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Retored it.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 9 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 10 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py:151]: Right. The `mask` arg used to carry the causal mask, but that's now handled by `create_causal_mask` and passed through `attention_mask`, so it's redundant.

File: src/transformers/models/ctrl/modeling_ctrl.py (hunk 11 above)
Comment (YangKai0616) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Done.

**Comment by YangKai0616** [comment]
> Have left a few smaller comments. We could further refactor to also have the same namings by using the conversion mapping but tbh, I don't think it's worth it for this model as it's very old and likely to be deprecated at some point

Agreed. Going forward, I'll prioritize model importance when deciding whether a model needs refactoring. Thanks for reviewing this PR!

**Review by vasqu — APPROVED** [review]
LGTM, checking with slow CI then merging 🫡

**Comment by vasqu** [comment]
run-slow: ctrl
